Use ~/.hatterrc if found.

ZachMassia opened the pull request:
Hatter will attempt to use ~/.hatterrc if found.

I've also tried cleaning up the configuration spec to be a little more DRY. Any style comments appreciated. :)

ZachMassia committed:
Use ~/.hatterrc if found

ZachMassia commented on spec/configuration_spec.rb:
```diff
@@ -4,43 +4,64 @@
+    end
+
+    it "should create a default config" do
+      File.exists?(config.default_config).should be_true
```
Kind of hacky, but because the Configuration class is a Singleton, there is no way to make another call to initialize.

ZachMassia commented on spec/configuration_spec.rb:
```diff
@@ -4,43 +4,64 @@
+    it "barfs when required values are missing" do
+      config_file.gsub!(/maildir/, 'asdf')
+      File.open("invalid_config", "w") {|file| file.write config_file }
+      expect {Configuration.instance "invalid_config"}.to raise_error
```
I believe this is raising an error for the wrong reason. 

Singleton.instance does not take any parameters, which is the error. It's not related to an invalid config file.

expez replied:
You are right, I messed this up while refactoring.  Nice catch!

It should be: Configuration.send(:new, "invalid_config") 

It has to look like this because Configuration is a singleton and thus has a private constructor.

ZachMassia replied:
Ah, neat. 

I tried searching quickly for some way of reinitializing a singleton but I didn't come up with much.

expez commented:
The idea is that the hatterrc file, included in the repo, will be copied into to /etc when the gem gets installed.  The user will then override the sitewide config file with a file in ~/.hatterrc.

Given that, I think it would better if you changed this code to check for a hatterc in ~/ to use, as the current code is only going to be helpful in a development context.

PS: I'm very impressed you managed to submit this PR only hours after deciding to work on Hatter!

ZachMassia commented:
Is copying hatterrc to /etc something that needs to be added as well?

expez commented:
I haven't looked into how `gem install` works, but that's the mechanism to use to copy that file into /etc.  Otherwise we would have to convince our users to give hatter root access and that might be a an uphill battle.

ZachMassia commented:
I can't remember if it's possible to do it when merging the pull request, but if needed I can squash those commits into one.

I'll read up on the whole gem packaging/distribution in the next few days and see what I can find.

ZachMassia commented:
Thanks for all the feedback. :)

I didn't even think of the ternary operator, coming from C++ it's not widely used. 

As for the invalid config bug, I believe you weren't looking at the latest commit, as I did fix that.

I think I had my DSL style mixed up with regular Ruby. For some reason I thought rubyists didn't like parens.

ZachMassia commented:
So I've fixed up all the other stuff, but I'm a little stuck on testing the config file. 

The following test only works if they have a ~/.hatterrc.

``` ruby
context 'with a user config file should override defaults' do
  its(:maildir)        { should_not eq maildir_path }
  its(:maildir_format) { should_not eq maildir_format }
end
```

In the case they don't have one, how would I go about testing?

edit: Also, thanks for the heads up on Rubocop. Looks like they have a ST2 plugin for it as well.

ZachMassia commented:
Ok, I've squashed all commits into one, and just added a basic test for now:

``` ruby
its(:maildir) { should_not eq maildir_path if user_file_exists }
# Same for maildir_format
```

I'm sure we can work out more exhaustive tests but at least for now we can use ~/.hatterrc.

expez commented:
$ rspec spec  
..............WARNING: error setting asdf with value /path/to/asdf: no parameter asdf exists for configuration class Configuration.  Ignoring unknown parameter asdf!
WARNING: error setting asdf_format with value asdf: no parameter asdf_format exists for configuration class Configuration.  Ignoring unknown parameter asdf_format!
........*..

Pending:
  Hatter should run tests
    # add some real tests
    # ./spec/hatter_spec.rb:4

Finished in 0.04244 seconds
25 examples, 0 failures, 1 pending

Do you get this output for the specs?  That warning is pretty weird.

Looks good otherwise!

ZachMassia commented:
Yep, same thing here. Looks like it's from the exception `configtoolkit` raises on the invalid config.

I did a quick search on Google about suppressing that but I didn't come up with anything.
